staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go

```diff
@@ -489,3 +489,423 @@ func TestCelCompilation(t *testing.T) {
+			expectedSetCost:  14,
+		},
+		{
+			name:             "map of strings with has",
+			schemaGenerator:  genMapWithRule("string", "has(self.x)"),
+			expectedCalcCost: 0,
+			setMaxLength:     550,
+			expectedSetCost:  0,
+		},
+		{
+			name:             "map of booleans with exists",
+			schemaGenerator:  genMapWithRule("boolean", "self.exists(x, true)"),
+			expectedCalcCost: 1874997,
+			setMaxLength:     25,
+			expectedSetCost:  127,
+		},
+		{
+			name:             "map of booleans with all",
+			schemaGenerator:  genMapWithRule("boolean", "self.all(x, true)"),
+			expectedCalcCost: 1499998,
+			setMaxLength:     100,
+			expectedSetCost:  402,
+		},
+		{
+			name:             "map of booleans with has",
+			schemaGenerator:  genMapWithRule("boolean", "has(self.x)"),
+			expectedCalcCost: 0,
+			setMaxLength:     1024,
+			expectedSetCost:  0,
+		},
+		{
+			name:             "string with contains",
+			schemaGenerator:  genStringWithRule("self.contains(\"test\")"),
+			expectedCalcCost: 300001,
+			setMaxLength:     10,
+			// TODO(DangerOnTheRanger): is this the expected cost in this scenario?
+			expectedSetCost: 2,
+		},
+		{
+			name:            "string with startsWith",
+			schemaGenerator: genStringWithRule("self.startsWith(\"test\")"),
+			// TODO(DangerOnTheRanger): is this the expected cost in this scenario?
+			expectedCalcCost: 2,
+			setMaxLength:     15,
+			expectedSetCost:  2,
+		},
+		{
+			name:            "string with endsWith",
+			schemaGenerator: genStringWithRule("self.endsWith(\"test\")"),
+			// TODO(DangerOnTheRanger): is this the expected cost in this scenario?
+			expectedCalcCost: 2,
+			setMaxLength:     30,
+			expectedSetCost:  2,
+		},
+		{
+			name:             "index of array with numbers",
+			schemaGenerator:  genArrayWithRule("number", "self[1] == 0.0"),
+			expectedCalcCost: 3,
+			setMaxLength:     5000,
+			expectedSetCost:  3,
+		},
+		{
+			name:            "index of array with strings",
+			schemaGenerator: genArrayWithRule("string", "self[1] == \"test\""),
+			// TODO(DangerOnTheRanger): is this the expected cost for this equality check?
+			expectedCalcCost: 3,
+			setMaxLength:     8,
+			expectedSetCost:  3,
+		},
+		{
+			name:            "O(n^2) loop with numbers",
+			schemaGenerator: genArrayWithRule("number", "self.all(x, self.all(y, true))"),
+			// TODO(DangerOnTheRanger): verify calculation here
+			expectedCalcCost: 8999997000000,
+			setMaxLength:     10,
+			expectedSetCost:  462,
+		},
+		{
+			name:             "O(n^3) loop with numbers",
+			schemaGenerator:  genArrayWithRule("number", "self.all(x, self.all(y, self.all(z, true)))"),
+			expectedCalcCost: 13499986500008999998,
+			setMaxLength:     10,
+			expectedSetCost:  4662,
+		},
+		{
+			name:             "regex matches simple",
+			schemaGenerator:  genStringWithRule("self.matches(\"x\")"),
+			expectedCalcCost: 300001,
+			setMaxLength:     50,
+			// TODO(DangerOnTheRanger): is this the expected cost here?
```

Comment:
Yes, this is right. the calculation is O(n*m)*costFactor where n is the input string len and m is the regex string len. The regex string length is short and the string is pretty short too.

Reply:
Gotcha, I've removed the TODO here.